### PR TITLE
Use `Sample` type with the RingBuffer

### LINKED ISSUE #165: Use Sample type in Ringbuffer
### What's needed?

The `OrderedRingBuffer`s update method takes a timestamp and a value as parameters.
The default use case for this class is to store data coming from the logical meter, which sends data of type `Sample` which members are a timestamp and a value.

### Proposed solution

Move he ringbuffer to `timeseries` and use the `Sample` class.

### Use cases

_No response_

### Alternatives and workarounds

Move the `Sample` class from the `timestamp` path to e.g. `util` and use it in the RingBuffer.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mathias-baumann-frequenz**
fixes #165

### CODE REVIEW

File: benchmarks/timeseries/benchmark_ringbuffer.py [hunk 1]
```diff
@@ -117,25 +116,26 @@ def main() -> None:
...
+    Array: ..........
+    List:  ..........
     Filling 29 days and running average & mean on every day:
-    Array: 0.09842290654996759 seconds
-    List:  0.1316629376997298 seconds
-    Diff:  -0.03324003114976222
+            Array: 7.6797407130012285 seconds
+            List:  8.115826181398006 seconds
+            Diff:  -0.43608546839677764
+
```
Comment (shsms) [inline benchmarks/timeseries/benchmark_ringbuffer.py]: Why do you think this is?  I can't believe creating a new `Sample` object for every message is making the benchmark 70-80x longer.

Maybe we should also try benchmarking the filling times and average/mean times separately?

File: benchmarks/timeseries/benchmark_ringbuffer.py [hunk 2]
```diff
@@ -30,12 +29,14 @@ def fill_buffer(
         # Push in random order
         for i in random.sample(range(MINUTES_IN_A_DAY), MINUTES_IN_A_DAY):
             buffer.update(
-                basetime + timedelta(days=day, minutes=i, seconds=i % 3),
-                element_type(i),
+                Sample(
+                    basetime + timedelta(days=day, minutes=i, seconds=i % 3),
+                    element_type(i),
```
Comment (shsms) [inline benchmarks/timeseries/benchmark_ringbuffer.py]: I guess only float will be supported from now on,  so we don't need the element_type.

File: benchmarks/timeseries/benchmark_ringbuffer.py (hunk 1 above)
Comment (mathias-baumann-frequenz) [reply benchmarks/timeseries/benchmark_ringbuffer.py]: Yah, the previous tests were faster due to missing window not down completely or properly I think, so I don't think we can compare that

File: src/frequenz/sdk/timeseries/_ringbuffer.py [hunk 3]
```diff
@@ -105,28 +106,25 @@ def maxlen(self) -> int:
         """
         return len(self._buffer)
 
-    def update(self, timestamp: datetime, value: T, missing: bool = False) -> None:
+    def update(self, sample: Sample) -> None:
```
Comment (matthias-wende-frequenz) [inline src/frequenz/sdk/timeseries/_ringbuffer.py:109]: Why are you removing the missing flag. That change seems unrelated.

Comment (matthias-wende-frequenz) [reply src/frequenz/sdk/timeseries/_ringbuffer.py:109]: Actually I liked that missing flag, because then the user can decide what's defined as missing. The NaN check is what we have in the `MovingWindow`.

File: src/frequenz/sdk/timeseries/_ringbuffer.py [hunk 4]
```diff
@@ -140,9 +138,10 @@ def update(self, timestamp: datetime, value: T, missing: bool = False) -> None:
         self._datetime_oldest = self._datetime_newest - self._time_range
 
         # Update data
+        value: float = math.nan if sample.value is None else sample.value
         self._buffer[self.datetime_to_index(timestamp)] = value
 
-        self._update_gaps(timestamp, prev_newest, missing)
+        self._update_gaps(timestamp, prev_newest, sample.value is None)
```
Comment (matthias-wende-frequenz) [inline src/frequenz/sdk/timeseries/_ringbuffer.py:144]: We should check if it's nan, since that's what the logical meter returns. The resampler though returns None. I think that's another argument to stick to the missing semantic.

**Review by matthias-wende-frequenz — COMMENTED** [review]
I feel we shouldn't remove the flag to indicate a missing sample.

File: src/frequenz/sdk/timeseries/_ringbuffer.py (hunk 3 above)
Comment (mathias-baumann-frequenz) [reply src/frequenz/sdk/timeseries/_ringbuffer.py:109]: Because `Sample` has an `optional`

File: src/frequenz/sdk/timeseries/_ringbuffer.py (hunk 4 above)
Comment (shsms) [reply src/frequenz/sdk/timeseries/_ringbuffer.py:144]: The LogicalMeter returns `None`, not `NaN`.  It uses `NaN`s just internally.

**Review by matthias-wende-frequenz — APPROVED** [review]
